File: solana_launchpads/launchpad_decoder/invariants.py
"""Self-consistency checks on decoded curve state.

A decoded price can be wrong in two very different ways. The layout can be
wrong, which `verify_against_node` catches. Or the *values* can be individually
plausible but mutually impossible -- a virtual reserve pair that does not lie on
the curve it claims to be on. That second kind survives every schema check and
shows up downstream as a wrong price on a coin that looks fine.

A constant-product curve gives a free test: ``vBase * vQuote`` is invariant.
The programs floor their outputs, so in practice k creeps *up* by parts per
billion over a curve's life and never moves down. A pair whose k differs from
the curve's opening k by a visible margin did not come off one account read.

The usual cause is field mixing rather than corruption -- two fields sourced
from different places and merged into one row. pump.fun makes this easy: its
`TradeEvent` carries `virtual_sol_reserves` (index 6) *and*
`virtual_quote_reserves` (index 30) as separate fields, the second appended
later for non-SOL quote pairs. Read one from the account and the other from an
event and the pair no longer multiplies to k.

The limit of this, stated plainly, because it was measured rather than guessed
and it is easy to overstate what a curve can tell you about itself.

The opening k is read from the program's config *now*. If a launchpad ever
changed its opening reserves, or seeds a different opening per quote mint,
curves created under another constant sit on a different k. Judged against
today's Global they are flagged, and they are not wrong -- the check is.
Measured against a real node on 60 sampled curves, every stored column held
exactly the on-chain field its name claimed and 59 of 60 still failed this
check. That is the check over-reaching, not the data being broken, and it was
rejecting roughly 9% of pump.fun curves.

So `check_curve_opening` asks the weaker question that survives dropping the
constants: is this opening *structurally possible*? And the honest answer to
"can a single row prove itself wrong without them?" is **no**. A curve's four
reserves determine its own opening exactly (see `CurveOpening`), so k against
that opening holds by construction -- the test is circular. What is left is
positivity, and nothing else.

Information about a curve that matches no known constant has to come from
outside the row, and there are three places it can come from:

* **the population** -- bucket `virtual_quote - real_quote` across all curves.
  A real opening is shared by thousands; a corrupted value is unique to its row.
* **the curve's history** -- k is conserved, so k must not *move* for a given
  curve between slots, whatever its value. This needs no constants at all and
  is the strongest test available.
* **the known constants** -- exact, but only for the openings the program uses
  today, which is where this started.

The first two are for an operator with the whole table, not for a decoder
holding one account, which is why this module reports a nonstandard opening as
a warning and leaves the verdict to whoever can see the population.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

#: k only ever creeps up, by parts per billion. 0.5% is far beyond any
#: legitimate drift while staying clear of floating-point noise.
K_TOLERANCE = 0.005

#: how far a virtual reserve may sit past its opening value before it is
#: treated as impossible rather than as rounding dust
OPEN_TOLERANCE = 1e-6


@dataclass(frozen=True)
class Violation:
    name: str
    detail: str
    severity: str = "error"

    def __str__(self) -> str:
        return f"{self.name}: {self.detail}"


@dataclass(frozen=True)
class CurveOpening:
    """A pump.fun curve's own opening parameters, recovered from its reserves.

    The program moves each virtual reserve in lockstep with its real
    counterpart -- a buy subtracts the same token amount from
    ``virtual_token_reserves`` and ``real_token_reserves``, and adds the same
    quote amount to ``virtual_quote_reserves`` and ``real_quote_reserves``.
    So two differences are fixed for the curve's whole life, and both of them
    are opening parameters:

        quote_seed = virtual_quote - real_quote  == initial_virtual_quote_reserves
        base_floor = virtual_base  - real_base   == initial_virtual_token_reserves
                                                    - initial_real_token_reserves

    `base_floor` is also the virtual base the curve ends on, which is why the
    graduation price is ``(quote_seed + raise) / base_floor``.

    Given those two, the opening k is ``(base_floor + initial_real_base) *
    quote_seed``, and since k is conserved that inverts to

        initial_real_base = virtual_base * virtual_quote / quote_seed - base_floor

    so a curve states its entire opening without the Global account.
    """

    quote_seed: int
    base_floor: int
    implied_initial_real_base: Optional[int]
    #: False when real_quote is zero -- the curve has not traded, so its
    #: reserves carry no evidence about k and self-consistency is vacuous.
    traded: bool

    @property
    def initial_virtual_base(self) -> Optional[int]:
        if self.implied_initial_real_base is None:
            return None
        return self.base_floor + self.implied_initial_real_base


def curve_opening(
    virtual_base: Optional[int],
    virtual_quote: Optional[int],
    real_base: Optional[int],
    real_quote: Optional[int],
) -> Optional[CurveOpening]:
    """Recover a curve's opening from its four reserve fields alone.

    Returns None when a field is missing. Note this needs all four: the two
    virtual reserves say where the curve is, the two real ones say how far it
    has come, and only together do they say where it started.
    """
    if virtual_base is None or virtual_quote is None:
        return None
    if real_base is None or real_quote is None:
        return None

    quote_seed = virtual_quote - real_quote
    base_floor = virtual_base - real_base
    implied_initial_real_base = None
    if quote_seed > 0:
        implied_initial_real_base = (virtual_base * virtual_quote) // quote_seed - base_floor
    return CurveOpening(
        quote_seed=quote_seed,
        base_floor=base_floor,
        implied_initial_real_base=implied_initial_real_base,
        traded=real_quote > 0,
    )


def check_curve_opening(
    opening: Optional[CurveOpening],
    token_total_supply: Optional[int] = None,
) -> List[Violation]:
    """Is a curve's recovered opening structurally possible?

    This is what survives when the program's *current* opening constants are
    not assumed. It asks only what must be true of any constant-product curve
    the program could have created, so it cannot reject a curve merely for
    having been seeded differently from today's default.

    That matters: a launchpad that changes its opening reserves, or seeds a
    different opening per quote mint, leaves a large population of curves that
    are entirely correct and match no current constant. Judging those against
    today's Global rejects real data.
    """
    violations: List[Violation] = []
    if opening is None:
        return violations

    if opening.quote_seed <= 0:
        violations.append(
            Violation(
                "quote_seed_not_positive",
                f"virtual_quote - real_quote is {opening.quote_seed}, but the "
                f"virtual quote reserve is seeded above zero and then tracks the "
                f"real one exactly, so the difference is positive for life. A "
                f"value at or below zero is what a virtual column carrying the "
                f"REAL reserve looks like",
            )
        )
    if opening.base_floor <= 0:
        violations.append(
            Violation(
                "base_floor_not_positive",
                f"virtual_base - real_base is {opening.base_floor}; the curve "
                f"keeps virtual base above real base for life, so this pair did "
                f"not come from one account read",
            )
        )

    implied = opening.implied_initial_real_base
    if implied is not None and opening.quote_seed > 0 and opening.base_floor > 0:
        if implied <= 0:
            violations.append(
                Violation(
                    "implied_opening_impossible",
                    f"the reserves imply the curve opened with {implied} sellable "
                    f"tokens, which cannot be",
                )
            )
        elif token_total_supply and implied > token_total_supply:
            violations.append(
                Violation(
                    "implied_opening_exceeds_supply",
                    f"the reserves imply the curve opened with {implied} sellable "
                    f"tokens against a total supply of {token_total_supply}",
                )
            )
    return violations


def classify_variant(
    virtual_quote: Optional[int],
    real_quote: Optional[int],
    candidates: Dict[str, int],
) -> Tuple[Optional[str], Optional[int]]:
    """Which opening constant a curve was seeded with, read off its reserves.

    pump.fun's buy and sell add the post-fee amount to the virtual *and* the
    real quote reserve together, so their difference never moves:

        virtual_quote - real_quote == initial_virtual_quote_reserves

    for the whole life of the curve. That makes the difference an exact
    classifier. It beats matching k against each candidate, because it needs no
    tolerance and -- unlike counting curves that still sit near their opening
    value -- it keeps working after the curve has traded away from it.

    Returns ``(variant name, its opening constant)``, or ``(None, observed
    offset)`` when the difference matches nothing, which means the pair did not
    come from one read.
    """
    if virtual_quote is None or real_quote is None:
        return None, None
    offset = virtual_quote - real_quote
    for name, initial in candidates.items():
        if initial and offset == initial:
            return name, initial
    return None, offset


def check_reserve_offset(
    virtual_quote: Optional[int],
    real_quote: Optional[int],
    candidates: Dict[str, int],
) -> List[Violation]:
    """Report a curve whose virtual/real quote difference matches no opening.

    This is a *warning*, not an error, and the distinction was learned the hard
    way. The difference is exactly the curve's opening quote reserve, so it is
    a reliable classifier -- but only against the openings the program is using
    today. A launchpad that has changed its seed, or that seeds per quote mint,
    leaves a large population of curves whose opening is real and simply not in
    the candidate list. Treating those as corrupt rejects correct data, and at
    scale it rejects a lot of it.

    An offset at or below zero is different in kind: no opening can be zero or
    negative, so that one is structurally impossible rather than merely
    unfamiliar. `check_curve_opening` raises it as an error.
    """
    variant, offset = classify_variant(virtual_quote, real_quote, candidates)
    if variant is not None or offset is None or offset <= 0:
        # offset <= 0 is check_curve_opening's to report, as an error.
        return []
    known = ", ".join(f"{name}={value}" for name, value in candidates.items() if value)
    return [
        Violation(
            "nonstandard_opening",
            f"virtual_quote - real_quote is {offset}, which matches no opening "
            f"constant the program uses today ({known}). That difference is the "
            f"curve's own opening quote reserve, fixed for its whole life, so "
            f"the curve is priced against {offset} here rather than against the "
            f"default. Whether that opening is real or the row is corrupt cannot "
            f"be told from this row: bucket the value across the population, "
            f"because a real opening is shared by many curves and a corrupt one "
            f"is unique to its row",
            severity="warning",
        )
    ]


def check_constant_product_pair(
    virtual_base: Optional[int],
    virtual_quote: Optional[int],
    initial_base: Optional[int],
    initial_quote: Optional[int],
    *,
    tolerance: float = K_TOLERANCE,
) -> List[Violation]:
    """Does this virtual reserve pair lie on the curve it claims to be on?

    All four arguments are raw on-chain units. Returns an empty list when the
    pair is consistent, which is the common case.
    """
    violations: List[Violation] = []
    if not virtual_base or not virtual_quote or not initial_base or not initial_quote:
        return violations

    k_open = initial_base * initial_quote
    k_now = virtual_base * virtual_quote
    drift = k_now / k_open - 1.0

    if abs(drift) > tolerance:
        violations.append(
            Violation(
                "k_violation",
                f"vBase*vQuote is {k_now:.6e}, the curve opened at {k_open:.6e} "
                f"({drift:+.1%}). Flooring moves k up by parts per billion and "
                f"never down, so this pair did not come from one account read",
            )
        )

    if virtual_quote < initial_quote * (1 - OPEN_TOLERANCE):
        violations.append(
            Violation(
                "quote_below_open",
                f"virtual quote {virtual_quote} is below the opening "
                f"{initial_quote}; a one-way curve cannot go below its seed",
            )
        )
    if virtual_base > initial_base * (1 + OPEN_TOLERANCE):
        violations.append(
            Violation(
                "base_above_open",
                f"virtual base {virtual_base} is above the opening {initial_base}; "
                f"more tokens would have been sold into the curve than ever left it",
            )
        )
    return violations


def diagnose_pair(
    virtual_base: Optional[int],
    virtual_quote: Optional[int],
    initial_base: Optional[int],
    initial_quote: Optional[int],
    base_for_sale: Optional[int] = None,
) -> Tuple[Optional[str], str]:
    """Given an inconsistent pair, say which side is the likelier culprit.

    Each field has a legal range on a one-way curve: the base reserve only ever
    falls from its opening (and not below what the curve may sell), the quote
    reserve only ever rises. Whichever observed value sits outside its range is
    the one to distrust, and the other implies what it should have been.

    Returns ``(suspect_field, explanation)``; `suspect_field` is None when the
    pair is consistent or cannot be judged.
    """
    if not virtual_base or not virtual_quote or not initial_base or not initial_quote:
        return None, "insufficient data"

    k_open = initial_base * initial_quote
    if abs(virtual_base * virtual_quote / k_open - 1.0) <= K_TOLERANCE:
        return None, "pair is consistent with the opening k"

    floor_base = initial_base - base_for_sale if base_for_sale else 0
    base_ok = floor_base <= virtual_base <= initial_base * (1 + OPEN_TOLERANCE)
    quote_ok = virtual_quote >= initial_quote * (1 - OPEN_TOLERANCE)

    implied_quote = k_open // virtual_base
    implied_base = k_open // virtual_quote

    if base_ok and not quote_ok:
        return (
            "virtual_quote",
            f"virtual base is in range, virtual quote is not; on this base the "
            f"quote should be {implied_quote} ({implied_quote / 1e9:.4f} "
            f"in 9-decimal units), not {virtual_quote}",
        )
    if quote_ok and not base_ok:
        return (
            "virtual_base",
            f"virtual quote is in range, virtual base is not; on this quote the "
            f"base should be {implied_base}, not {virtual_base}",
        )
    if not base_ok and not quote_ok:
        # Both are outside their range, but rarely by the same margin. The one
        # that has travelled furthest past its bound is the one to chase first.
        base_excursion = max(0.0, (virtual_base - initial_base) / initial_base)
        quote_excursion = max(0.0, (initial_quote - virtual_quote) / initial_quote)
        if quote_excursion >= base_excursion:
            return (
                "virtual_quote",
                f"both reserves are out of range, but the quote is further out "
                f"({quote_excursion:.1%} below its opening vs {base_excursion:.1%} "
                f"above for the base); on this base the quote should be "
                f"{implied_quote} ({implied_quote / 1e9:.4f} in 9-decimal units), "
                f"not {virtual_quote}",
            )
        return (
            "virtual_base",
            f"both reserves are out of range, but the base is further out "
            f"({base_excursion:.1%} above its opening vs {quote_excursion:.1%} "
            f"below for the quote); on this quote the base should be "
            f"{implied_base}, not {virtual_base}",
        )
    return None, "both reserves are individually in range but their product is not k"


def check_metrics(metrics, *, tolerance: float = K_TOLERANCE) -> List[Violation]:
    """Run whatever invariants apply to an already-decoded `LaunchMetrics`.

    Price-level rather than reserve-level, so it works for any curve family:
    the launch price is the floor a one-way curve can never trade below, and
    the graduation price is the ceiling it cannot pass while still on the curve.
    """
    from .types import CurveFamily

    violations: List[Violation] = []
    launch = metrics.launch_price_quote.value
    now = metrics.current_price_quote.value
    graduation = metrics.graduation_price_quote.value

    if launch is not None and now is not None and launch > 0:
        if now < launch * (1 - OPEN_TOLERANCE):
            violations.append(
                Violation(
                    "price_below_launch",
                    f"current price {now:.6e} is below the launch price {launch:.6e}; "
                    f"a one-way curve cannot trade below its opening",
                )
            )
    if (
        graduation is not None
        and now is not None
        and graduation > 0
        and metrics.complete.value is False
        and now > graduation * (1 + OPEN_TOLERANCE)
    ):
        violations.append(
            Violation(
                "price_above_graduation",
                f"current price {now:.6e} exceeds the graduation price "
                f"{graduation:.6e} while the curve is still marked incomplete",
            )
        )

    raised = metrics.raised_quote.value
    target = metrics.raise_target_quote.value
    if raised is not None and target and raised > target * 1.02:
        violations.append(
            Violation(
                "raised_above_target",
                f"raised {raised} exceeds the raise target {target}",
                severity="warning",
            )
        )

    if metrics.curve_family is CurveFamily.CONSTANT_PRODUCT_VIRTUAL:
        state = metrics.raw_state or {}
        quote = state.get("virtual_quote_reserves")
        if quote is None:
            quote = state.get("virtual_sol_reserves")
        violations.extend(
            check_constant_product_pair(
                state.get("virtual_token_reserves"),
                quote,
                state.get("_initial_virtual_base"),
                state.get("_initial_virtual_quote"),
                tolerance=tolerance,
            )
        )
    return violations

File: solana_launchpads/go/launchpad/invariants.go
package launchpad

import (
	"fmt"
	"math/big"
	"math/bits"
)

// Self-consistency checks on decoded curve state.
//
// A decoded price can be wrong in two ways. The layout can be wrong, which the
// Python-side selftest catches against a live cluster. Or the values can be
// individually plausible but mutually impossible -- a reserve pair that does
// not lie on the curve it claims to be on. The second kind passes every schema
// check and shows up downstream as a wrong price on a coin that looks fine.
//
// A constant-product curve gives the test for free: vBase*vQuote is invariant.
// Programs floor their outputs, so k creeps *up* by parts per billion over a
// curve's life and never moves down. A pair missing k by a visible margin did
// not come off one account read -- in practice, two fields sourced from
// different places and merged into one row.
//
// One limitation, stated plainly: the opening k comes from the program's config
// as it stands now. If a launchpad ever changed its opening reserves, curves
// created before the change sit on a different k and will be flagged although
// they are correct. That false positive is a cluster of old curves all off by
// the same ratio; mixed fields scatter. Bucketing the ratio tells them apart.

const (
	// KTolerance is far beyond any legitimate drift while staying clear of
	// floating-point noise.
	KTolerance = 0.005
	// OpenTolerance is how far a reserve may sit past its opening value before
	// it is treated as impossible rather than as rounding dust.
	OpenTolerance = 1e-6
)

// Severity distinguishes a wrong number from a stale one.
type Severity string

const (
	SeverityError   Severity = "error"
	SeverityWarning Severity = "warning"
)

// Violation is one broken invariant.
type Violation struct {
	Name     string
	Detail   string
	Severity Severity
}

func (v Violation) String() string { return v.Name + ": " + v.Detail }

// CheckConstantProductPair reports whether a virtual reserve pair lies on the
// curve it claims to be on. Returns nil for a consistent pair, which is the
// common case and allocates nothing.
func CheckConstantProductPair(virtualBase, virtualQuote, initialBase, initialQuote uint64) []Violation {
	if virtualBase == 0 || virtualQuote == 0 || initialBase == 0 || initialQuote == 0 {
		return nil
	}

	kOpen := MulU64(initialBase, initialQuote).Float()
	kNow := MulU64(virtualBase, virtualQuote).Float()
	drift := kNow/kOpen - 1

	var out []Violation
	if drift > KTolerance || drift < -KTolerance {
		out = append(out, Violation{
			Name: "k_violation",
			Detail: fmt.Sprintf(
				"vBase*vQuote is %.6e, the curve opened at %.6e (%+.1f%%). Flooring "+
					"moves k up by parts per billion and never down, so this pair did "+
					"not come from one account read", kNow, kOpen, drift*100),
			Severity: SeverityError,
		})
	}
	if float64(virtualQuote) < float64(initialQuote)*(1-OpenTolerance) {
		out = append(out, Violation{
			Name: "quote_below_open",
			Detail: fmt.Sprintf("virtual quote %d is below the opening %d; a one-way "+
				"curve cannot go below its seed", virtualQuote, initialQuote),
			Severity: SeverityError,
		})
	}
	if float64(virtualBase) > float64(initialBase)*(1+OpenTolerance) {
		out = append(out, Violation{
			Name: "base_above_open",
			Detail: fmt.Sprintf("virtual base %d is above the opening %d; more tokens "+
				"would have been sold into the curve than ever left it", virtualBase, initialBase),
			Severity: SeverityError,
		})
	}
	return out
}

// DiagnosePair names the likelier culprit in an inconsistent pair.
//
// Each reserve has a legal range on a one-way curve: the base only ever falls
// from its opening, the quote only ever rises. Whichever observed value sits
// outside its range is the one to distrust, and the other implies what it
// should have been. When both are out, the one that has travelled further past
// its bound is named first.
//
// Returns an empty field name when the pair is consistent or cannot be judged.
func DiagnosePair(virtualBase, virtualQuote, initialBase, initialQuote, baseForSale uint64) (string, string) {
	if virtualBase == 0 || virtualQuote == 0 || initialBase == 0 || initialQuote == 0 {
		return "", "insufficient data"
	}

	kOpen := MulU64(initialBase, initialQuote)
	drift := MulU64(virtualBase, virtualQuote).Float()/kOpen.Float() - 1
	if drift <= KTolerance && drift >= -KTolerance {
		return "", "pair is consistent with the opening k"
	}

	impliedQuote := divU128ByU64(kOpen, virtualBase)
	impliedBase := divU128ByU64(kOpen, virtualQuote)

	var floorBase uint64
	if baseForSale > 0 && baseForSale < initialBase {
		floorBase = initialBase - baseForSale
	}
	baseOK := virtualBase >= floorBase && float64(virtualBase) <= float64(initialBase)*(1+OpenTolerance)
	quoteOK := float64(virtualQuote) >= float64(initialQuote)*(1-OpenTolerance)

	switch {
	case baseOK && !quoteOK:
		return "virtual_quote", fmt.Sprintf(
			"virtual base is in range, virtual quote is not; on this base the quote "+
				"should be %d (%.4f in 9-decimal units), not %d",
			impliedQuote, float64(impliedQuote)/1e9, virtualQuote)
	case quoteOK && !baseOK:
		return "virtual_base", fmt.Sprintf(
			"virtual quote is in range, virtual base is not; on this quote the base "+
				"should be %d, not %d", impliedBase, virtualBase)
	case !baseOK && !quoteOK:
		baseExcursion := 0.0
		if virtualBase > initialBase {
			baseExcursion = float64(virtualBase-initialBase) / float64(initialBase)
		}
		quoteExcursion := 0.0
		if virtualQuote < initialQuote {
			quoteExcursion = float64(initialQuote-virtualQuote) / float64(initialQuote)
		}
		if quoteExcursion >= baseExcursion {
			return "virtual_quote", fmt.Sprintf(
				"both reserves are out of range, but the quote is further out (%.1f%% "+
					"below its opening vs %.1f%% above for the base); on this base the "+
					"quote should be %d (%.4f in 9-decimal units), not %d",
				quoteExcursion*100, baseExcursion*100, impliedQuote,
				float64(impliedQuote)/1e9, virtualQuote)
		}
		return "virtual_base", fmt.Sprintf(
			"both reserves are out of range, but the base is further out (%.1f%% above "+
				"its opening vs %.1f%% below for the quote); on this quote the base "+
				"should be %d, not %d",
			baseExcursion*100, quoteExcursion*100, impliedBase, virtualBase)
	}
	return "", "both reserves are individually in range but their product is not k"
}

// divU128ByU64 floors a 128-bit value by a 64-bit divisor.
//
// The two fast paths cover every curve in practice and allocate nothing, which
// matters because recovering a curve's opening runs on the hot path. bits.Div64
// panics when the quotient would not fit, so it is guarded by the exact
// condition that makes it safe; the big.Int fallback is for values that cannot
// arise from a real curve but must not panic if they do.
func divU128ByU64(value U128, divisor uint64) uint64 {
	if divisor == 0 {
		return 0
	}
	if value.Hi == 0 {
		return value.Lo / divisor
	}
	if value.Hi < divisor {
		quotient, _ := bits.Div64(value.Hi, value.Lo, divisor)
		return quotient
	}
	n := new(big.Int).Lsh(new(big.Int).SetUint64(value.Hi), 64)
	n.Or(n, new(big.Int).SetUint64(value.Lo))
	n.Div(n, new(big.Int).SetUint64(divisor))
	if !n.IsUint64() {
		return 0
	}
	return n.Uint64()
}

// VariantCandidates maps a variant name to the opening quote reserve pump.fun
// seeds that variant with. SOL-quoted and non-SOL-quoted curves get different
// openings and the account does not say which it is.
type VariantCandidates map[string]uint64

// ClassifyVariant reports which opening constant a curve was seeded with, read
// off the reserves themselves.
//
// pump.fun's buy and sell add the post-fee amount to the virtual *and* the real
// quote reserve together, so their difference never moves:
//
//	virtual_quote - real_quote == initial_virtual_quote_reserves
//
// for the whole life of the curve. That makes the difference an exact
// classifier: no tolerance, and it keeps working after a curve has traded well
// away from its opening, which is exactly where counting curves still sitting
// near that opening fails.
//
// Returns the variant name and its constant, or ("", observed offset) when the
// difference matches nothing -- which means the pair did not come from one read.
func ClassifyVariant(virtualQuote, realQuote uint64, candidates VariantCandidates) (string, int64) {
	offset := int64(virtualQuote) - int64(realQuote)
	for name, initial := range candidates {
		if initial != 0 && offset == int64(initial) {
			return name, offset
		}
	}
	return "", offset
}

// CheckReserveOffset reports a curve whose virtual/real quote difference
// matches no opening constant.
//
// This is a WARNING, not an error, and the distinction was learned from field
// data. The difference is exactly the curve's opening quote reserve, which
// makes it an exact classifier -- but only against the openings the program
// uses today. A launchpad that has changed its seed, or seeds one per quote
// mint, leaves a large population of curves whose opening is real and simply
// not in the candidate list. Measured against a node on 60 sampled curves,
// every stored column held exactly the on-chain field its name claimed and 59
// still failed the old error-level check. That was the check over-reaching, on
// roughly 9% of pump.fun curves.
//
// An offset at or below zero is different in kind: no curve opens at zero, so
// that is impossible rather than unfamiliar, and CheckCurveOpening raises it
// as an error.
func CheckReserveOffset(virtualQuote, realQuote uint64, candidates VariantCandidates) []Violation {
	if len(candidates) == 0 {
		return nil
	}
	variant, offset := ClassifyVariant(virtualQuote, realQuote, candidates)
	if variant != "" || offset <= 0 {
		return nil
	}
	return []Violation{{
		Name: "nonstandard_opening",
		Detail: fmt.Sprintf(
			"virtual_quote - real_quote is %d, which matches no opening constant "+
				"the program uses today. That difference is the curve's own opening "+
				"quote reserve, fixed for its whole life, so the curve is priced "+
				"against %d here rather than against the default. Whether that "+
				"opening is real or the row is corrupt cannot be told from this row: "+
				"bucket the value across the population, because a real opening is "+
				"shared by many curves and a corrupt one is unique to its row",
			offset, offset),
		Severity: SeverityWarning,
	}}
}

// CurveOpening is a pump.fun curve's own opening parameters, recovered from its
// four reserve fields.
//
// The program moves each virtual reserve in lockstep with its real
// counterpart, so two differences are fixed for the curve's whole life and both
// of them are opening parameters:
//
//	QuoteSeed = virtual_quote - real_quote == initial_virtual_quote_reserves
//	BaseFloor = virtual_base  - real_base  == initial_virtual_token_reserves
//	                                          - initial_real_token_reserves
//
// BaseFloor is also the virtual base the curve ends on, which is why the
// graduation price is (QuoteSeed + raise) / BaseFloor.
//
// Since k is conserved, the opening inverts out of those two:
//
//	initial_real_base = virtual_base * virtual_quote / QuoteSeed - BaseFloor
//
// So a curve states its entire opening without the Global account -- and, note,
// k against that opening then holds by construction. A single row cannot prove
// itself wrong once the program's constants are not assumed; only positivity is
// left. The evidence has to come from the population (bucket QuoteSeed across
// all curves: a real opening is shared by thousands) or from the curve's own
// history (k must not move between slots).
type CurveOpening struct {
	QuoteSeed int64
	BaseFloor int64
	// ImpliedInitialRealBase is zero when QuoteSeed is not positive.
	ImpliedInitialRealBase uint64
	// Traded is false when real_quote is zero: the curve carries no evidence
	// about k at all, so self-consistency is doubly vacuous.
	Traded bool
	// OK is false when a field was missing and nothing could be recovered.
	OK bool
}

// InitialVirtualBase is the virtual base the curve opened with.
func (o CurveOpening) InitialVirtualBase() uint64 {
	if !o.OK || o.BaseFloor <= 0 {
		return 0
	}
	return uint64(o.BaseFloor) + o.ImpliedInitialRealBase
}

// RecoverOpening derives a curve's opening from its four reserves. It needs all
// four: the virtual pair says where the curve is, the real pair says how far it
// has come, and only together do they say where it started.
func RecoverOpening(virtualBase, virtualQuote, realBase, realQuote uint64) CurveOpening {
	o := CurveOpening{
		QuoteSeed: int64(virtualQuote) - int64(realQuote),
		BaseFloor: int64(virtualBase) - int64(realBase),
		Traded:    realQuote > 0,
		OK:        true,
	}
	if o.QuoteSeed > 0 {
		k := MulU64(virtualBase, virtualQuote)
		implied := int64(divU128ByU64(k, uint64(o.QuoteSeed))) - o.BaseFloor
		if implied > 0 {
			o.ImpliedInitialRealBase = uint64(implied)
		}
	}
	return o
}

// CheckCurveOpening asks the one question that survives dropping the program's
// constants: is this opening structurally possible?
//
// It cannot reject a curve merely for having been seeded differently from
// today's default, which is the whole point -- that rejection was wrong, and at
// scale it was wrong about a lot of correct data.
func CheckCurveOpening(o CurveOpening, tokenTotalSupply uint64) []Violation {
	if !o.OK {
		return nil
	}
	var out []Violation
	if o.QuoteSeed <= 0 {
		out = append(out, Violation{
			Name: "quote_seed_not_positive",
			Detail: fmt.Sprintf(
				"virtual_quote - real_quote is %d, but the virtual quote reserve is "+
					"seeded above zero and then tracks the real one exactly, so the "+
					"difference is positive for life. A value at or below zero is what "+
					"a virtual column carrying the REAL reserve looks like", o.QuoteSeed),
			Severity: SeverityError,
		})
	}
	if o.BaseFloor <= 0 {
		out = append(out, Violation{
			Name: "base_floor_not_positive",
			Detail: fmt.Sprintf(
				"virtual_base - real_base is %d; the curve keeps virtual base above "+
					"real base for life, so this pair did not come from one account read",
				o.BaseFloor),
			Severity: SeverityError,
		})
	}
	if o.QuoteSeed > 0 && o.BaseFloor > 0 {
		implied := o.ImpliedInitialRealBase
		switch {
		case implied == 0:
			out = append(out, Violation{
				Name:     "implied_opening_impossible",
				Detail:   "the reserves imply the curve opened with no sellable tokens",
				Severity: SeverityError,
			})
		case tokenTotalSupply > 0 && implied > tokenTotalSupply:
			out = append(out, Violation{
				Name: "implied_opening_exceeds_supply",
				Detail: fmt.Sprintf(
					"the reserves imply the curve opened with %d sellable tokens "+
						"against a total supply of %d", implied, tokenTotalSupply),
				Severity: SeverityError,
			})
		}
	}
	return out
}
